zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1165,20 +1177,92 @@ private void reconnect() {
     }
   }
 
+  private class SyncContext {
```

Comment:
Please add the callback logic to ZkAsyncCallbacks.java if possible.
1. This file has been very large. We shall avoid adding more optional content here.
2. ZkAsyncCallbacks.java has all the callback defines there. We might be able to avoid some duplicated code if the class is defined there and extending the default callback class.

Reply:
Before this pull request, I already had a look. THe reason not to use ZkAsyncCallbacks are two fold:
1/ Upon failure, the syncContext need to block the current ZkEvent queue. ZkAsyncCallbacks does not block the ZkEvent queue or the _asycThread event queue, but retry from _asyncThread. ---- Thread modeling difference.
2/ ZkAsyncCallbacks upon session expiration, would still retry. For this SyncContext, it did not retry session expiration. -- Error handling difference.

Especially for thread modeling difference, it is hard to merge them together.

Reply:
I might miss some points here. But if we call DefaultCallback.waitForSuccess() in the new session handling method at where you call the latch.await(), will it do the same trick?

Regarding the retry logic, it is simple to extend the DefaultCallback.needRetry() in the SyncCallbackHandler.
1. make needRetry() protected.
2. override it in the new SyncCallbackHandler class.